# SQL

# Задания

1) Посчитать, сколько книг вышло после 1 января 2000 года;

2) Для каждой книги посчитать количество обзоров и среднюю оценку;

3) Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;

4) Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;

5) Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


# Описание данных

    Таблица books cодержит данные о книгах:
    
book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

    Таблица authors cодержит данные об авторах:
    
author_id — идентификатор автора;

author — имя автора.

    Таблица publishers содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

Таблица ratings содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

    Таблица reviews содержит данные о пользовательских обзорах на книги:
    
review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя пользователя, написавшего обзор;

text — текст обзора.

## Загрузка и исследование данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': '', # имя пользователя
'pwd': '', # пароль
'host': '',
'port': , # порт подключения
'db': ''} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
tables  = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table in tables:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    for_table = pd.io.sql.read_sql(query, con = engine) 
    print(f'Таблица "{table}"')
    display(for_table.head())
    print(for_table.info())

Таблица "books"


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица "authors"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица "publishers"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица "ratings"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица "reviews"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None


### Вывод

В данных отсутствуют пропуски, можно заменить тип данных в таблице books колонка publication_date. Но для написания запросов это не требуется. Можно приступать к написанию запросов.

## Задание 1 

Посчитать, сколько книг вышло после 1 января 2000 года.

In [4]:
query1 = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01';
'''
request1 = pd.io.sql.read_sql(query1, con = engine)
book_count = request1.loc[0, 'count']
print(f'После 1 января 2000 года вышло {book_count} книг')

После 1 января 2000 года вышло 819 книг


### Вывод 

После 1 января 2000 года вышло 819 книг.

## Задание 2

Для каждой книги посчитать количество обзоров и среднюю оценку.

In [6]:
query2 = ''' 

SELECT 
    b.book_id,
    b.title,
    re.count_rewiew, 
    ra.avg_rating
FROM books AS b
LEFT JOIN  (SELECT book_id,
                   AVG(rating) AS avg_rating
            FROM ratings
            GROUP BY(book_id)) AS ra ON b.book_id= ra.book_id
LEFT JOIN (SELECT book_id,
                  COUNT(review_id) AS count_rewiew
           FROM reviews
           GROUP BY(book_id)) AS re ON b.book_id = re.book_id
ORDER BY  re.count_rewiew DESC,
          re.count_rewiew DESC;  

'''
request2 = pd.io.sql.read_sql(query2, con = engine)
request2.head(10)

book_id                                              title  count_rewiew  \
0      808                            The Natural Way to Draw           NaN   
1      191  Disney's Beauty and the Beast (A Little Golden...           NaN   
2      221                          Essential Tales and Poems           NaN   
3       83    Anne Rice's The Vampire Lestat: A Graphic Novel           NaN   
4      387                               Leonardo's Notebooks           NaN   
5      672   The Cat in the Hat and Other Dr. Seuss Favorites           NaN   
6      948                            Twilight (Twilight  #1)           7.0   
7      302  Harry Potter and the Prisoner of Azkaban (Harr...           6.0   
8      779  The Lightning Thief (Percy Jackson and the Oly...           6.0   
9      695  The Curious Incident of the Dog in the Night-Time           6.0   

   avg_rating  
0    3.000000  
1    4.000000  
2    4.000000  
3    3.666667  
4    4.000000  
5    5.000000  
6    3.662500  
7    4.414634  
8    4.080645  
9    4.081081

### Вывод 

Самое большое количество обзоров (7) у книги Twilight (Twilight #1) и самая высокая оценка - 5.

## Задание 3

Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.

In [7]:
query3 = ''' 

SELECT b.publisher_id,
       p.publisher,
       COUNT(b.book_id)
FROM books AS b 
LEFT JOIN publishers as p ON b.publisher_id=p.publisher_id
WHERE num_pages>50
GROUP BY
        b.publisher_id,
        p.publisher
ORDER BY COUNT(b.book_id)DESC
Limit 1; 
''' 

request3 = pd.io.sql.read_sql(query3, con = engine)
publisher_max = request3.loc[0, 'publisher']
books_max = request3.loc[0, 'count']
print(f'Издательство, которое выпустило наибольшее число книг ({books_max} штуки) толще 50 страниц - {publisher_max}.')

Издательство, которое выпустило наибольшее число книг (42 штуки) толще 50 страниц - Penguin Books.


### Вывод 

Издательство, которое выпустило наибольшее число книг (42 штуки) толще 50 страниц - Penguin Books.

## Задание 4

Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками.

In [9]:
query4 = ''' 
SELECT first.author,
       AVG(book_avg) AS author_AVG
FROM (SELECT r.book_id,
                    r.book_avg,
                    r.count_rating,
                    a.author
             FROM (SELECT book_id,
                   COUNT(rating_id) AS count_rating,
                   AVG(rating) AS book_avg 
                 FROM ratings 
                 GROUP BY book_id
                 HAVING COUNT(rating_id) >=50) AS r
             LEFT JOIN books AS b ON r.book_id = b.book_id
             LEFT JOIN authors AS a ON b.author_id = a.author_id) AS first
GROUP BY first.author
ORDER BY author_AVG DESC
Limit 1; 
'''

request4 = pd.io.sql.read_sql(query4, con = engine)
author_max = request4.loc[0, 'author']
author_avg_max4 = request4.loc[0, 'author_avg'].round(6)
print(f'Автор с самой высокой средней оценкой книг ({author_avg_max4}) - {author_max}.')

Автор с самой высокой средней оценкой книг (4.283844) - J.K. Rowling/Mary GrandPré.


### Вывод 

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré.

## Задание 5

Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
query5 = ''' 
SELECT AVG(first.count_review) AS avg_review
FROM (SELECT username,
             COUNT(review_id) AS count_review
      FROM reviews AS re
      WHERE username IN (
            SELECT username
            FROM ratings
            GROUP BY username
            HAVING COUNT(rating_id) >=50) 
      GROUP BY username) AS first
; 
''' 

request5 = pd.io.sql.read_sql(query5, con = engine)
review_max = request5.loc[0, 'avg_review'].round(1)
print(f'Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - {review_max}')

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.2


### Вывод 

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,2 .

## Вывод



После 1 января 2000 года вышло 819 книг.

Самое большое количество обзоров 7 у книги Twilight (Twilight #1) и самая высокая оценка 5.

Издательство, которое выпустило наибольшее число книг (42 штуки) толще 50 страниц - Penguin Books.

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré.

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,2 . 
